### PR TITLE
908: improve validate support

### PR DESCRIPTION
**Pull request by ademariag**
Fixes issue #908

## Proposed Changes

Overhaul of the validation support to greatly improve its use.
New proposed configuration:

```
    validate:
    - type: kubernetes
      fail_on_error: false
      verbose: true
      exclude:
        kinds:
        - HorizontalPodAutoscaler
        paths:
        - manifests/*-ingress.yml
      output_paths:
        - manifests/*.yml
      version: 1.26.0
 ```
 
## Example output
 ```
kapitan compile -t tesoro prod-sockshop argocd
Rendered inventory (0.18s)
Compiled tesoro (0.07s)
Compiled prod-sockshop (0.28s)
Compiled argocd (0.81s)
Validated tesoro FAIL
INVALID [MutatingWebhookConfiguration]: manifest ./compiled/tesoro/manifests/tesoro-bundle.yml
---> ERROR ['webhooks', 0]: 'sideEffects' is a required property
---> ERROR ['webhooks', 0]: 'admissionReviewVersions' is a required property

Validated prod-sockshop OK (with 3 exclusion/s)
Validated argocd OK (with 5 exclusion/s)
Validated targets (7.70s)
```
 

## Features
- supports globbed paths on output_paths
- uses https://github.com/kapicorp/kubernetes-json-schema instead of deprecated repo
- improved output
- only fails at the end of the validation to surface all errors at once
- autodetects resource kind
- silently ignore resources not available in the schema
- does not output anything if there are no files to validate (for non `kubernetes` targets)
- detects if schema exists by querying well known url for "all" kind
- supports files which contain multiple yaml documents
- support "fails_on_error" flag to make explicit whether to fail validation or not
- exclusion based on 
  - kind
  - file
  - annotation

### CODE REVIEW

File: kapitan/validator/kubernetes_validator.py [hunk 1]
```diff
@@ -2,6 +2,8 @@
 #
 # SPDX-License-Identifier: Apache-2.0
 
+import glob
```
Comment (ademariag) [inline kapitan/validator/kubernetes_validator.py:5]: needs removing

File: kapitan/validator/kubernetes_validator.py [hunk 2]
```diff
@@ -21,31 +23,36 @@ class KubernetesManifestValidator(Validator):
     def __init__(self, cache_dir, **kwargs):
         super().__init__(cache_dir, **kwargs)
 
-    def validate(self, validate_paths, **kwargs):
+    def validate(self, validate_files, **kwargs):
```
Comment (ademariag) [inline kapitan/validator/kubernetes_validator.py:26]: fix doc string

File: kapitan/validator/kubernetes_validator.py [hunk 3]
```diff
@@ -21,31 +21,113 @@ class KubernetesManifestValidator(Validator):
...
+            validate_files, errors, exclusions = self._validate_config_block(config)
+
+            if errors:
+                message += "FAIL"
+                error_message = ""
+                for error in errors:
+                    error_message += error
+
+                logger.info(message)
+                if config.get("fail_on_error", False):
```
Comment (ademariag) [inline kapitan/validator/kubernetes_validator.py:119]: default to true or flag parameter

**Comment by Moep90** [comment]
@ademariag we should make sure that some sort of globbing works as users may or may not have a nested directory structure.

1. `validate[*].exclude` and `validate[*].exclude.paths` are seamingly required even If I dont want to exclude stuff
2. I may or may not want to specify a `Kind`
3. If possible, can we make sure that `1.18` instead of `1.18.0` always uses the latest major version? In this case it would for example be `1.18.20`

## This....
```yaml
    validate: 
    - type: kubernetes 
      output_paths: 
        - "k8s/**/*.yml"
        - "k8s/**/*.yaml"
      exclude:
        paths: []
      version: 1.18.0
```

## leads to:
```
$ kapitan validate
Unknown (Non-Kapitan) Error occured
multiprocessing.pool.RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/home/linuxbrew/.linuxbrew/opt/python@3.10/lib/python3.10/multiprocessing/pool.py", line 125, in worker
    result = (True, func(*args, **kwds))
  File "/home/danny/.env3_kapitan_fix_validate/lib/python3.10/site-packages/kapitan-0.31.0rc0-py3.10.egg/kapitan/targets.py", line 856, in schema_validate_kubernetes_output
    KubernetesManifestValidator(cache_dir).validate(validate_data, **kwargs)
  File "/home/danny/.env3_kapitan_fix_validate/lib/python3.10/site-packages/kapitan-0.31.0rc0-py3.10.egg/kapitan/validator/kubernetes_validator.py", line 116, in validate
    validate_files, errors, exclusions = self._validate_config_block(config)
  File "/home/danny/.env3_kapitan_fix_validate/lib/python3.10/site-packages/kapitan-0.31.0rc0-py3.10.egg/kapitan/validator/kubernetes_validator.py", line 50, in _validate_config_block
    kind = validate_instance.get("kind")
AttributeError: 'NoneType' object has no attribute 'get'
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/danny/.env3_kapitan_fix_validate/lib/python3.10/site-packages/kapitan-0.31.0rc0-py3.10.egg/kapitan/targets.py", line 779, in schema_validate_compiled
    [p.get() for p in pool.imap_unordered(worker, validate_map.items()) if p]
  File "/home/danny/.env3_kapitan_fix_validate/lib/python3.10/site-packages/kapitan-0.31.0rc0-py3.10.egg/kapitan/targets.py", line 779, in <listcomp>
    [p.get() for p in pool.imap_unordered(worker, validate_map.items()) if p]
  File "/home/linuxbrew/.linuxbrew/opt/python@3.10/lib/python3.10/multiprocessing/pool.py", line 873, in next
    raise value
AttributeError: 'NoneType' object has no attribute 'get'


'NoneType' object has no attribute 'get'
```

# How to reproduce?
1. `git clone https://github.com/kapicorp/kapitan.git`
2. `python3 -m venv ~/.env3_kapitan_fix_validate`
3. `source ~/.env3_kapitan_fix_validate/bin/activate`
4. `python3 setup.py install`

**Pull request closed without merging**